src/zjs_gfx.c

```diff
@@ -0,0 +1,417 @@
+    gfx_data_t argData;
+    args_to_data(&argData, argc, argv);
+    return zjs_gfx_fill_rect_priv(argData.coords[0], argData.coords[1], argData.coords[2], argData.size, argData.color, handle);
+}
+
+static ZJS_DECL_FUNC(zjs_gfx_draw_rect)
+{
+    // requires: Requires 5 arguments
+    //           arg[0] - x coord of the top left.
+    //           arg[1] - y coord of the top left.
+    //           arg[2] - width.
+    //           arg[3] - height.
+    //           arg[4] - color array.
+    //           arg[5] - optional size of line.
+    //
+    //  effects: Draws a rectangle on the screen
+
+    ZJS_VALIDATE_ARGS(Z_NUMBER, Z_NUMBER, Z_NUMBER, Z_NUMBER, Z_ARRAY, Z_OPTIONAL Z_NUMBER);
+    ZJS_GET_HANDLE(this, gfx_handle_t, handle, gfx_type_info);
+    gfx_data_t argData;
+    args_to_data(&argData, argc, argv);
+    zjs_gfx_fill_rect_priv(argData.coords[0], argData.coords[1], argData.coords[2], argData.size, argData.color, handle);
+    zjs_gfx_fill_rect_priv(argData.coords[0], argData.coords[1] + argData.coords[3] - argData.size, argData.coords[2], argData.size, argData.color, handle);
+    zjs_gfx_fill_rect_priv(argData.coords[0], argData.coords[1], argData.size, argData.coords[3], argData.color, handle);
+    zjs_gfx_fill_rect_priv(argData.coords[0] + argData.coords[2] - argData.size, argData.coords[1], argData.size, argData.coords[3], argData.color, handle);
+
+    return ZJS_UNDEFINED;
+}
+
+static ZJS_DECL_FUNC(zjs_gfx_draw_char)
+{
+    // requires: Requires 4 arguments
+    //           arg[0] - x coord of the top left.
+    //           arg[1] - y coord of the top left.
+    //           arg[2] - character.
+    //           arg[3] - color array.
+    //           arg[4] - optional size of character.
+    //
+    //  effects: Draws a character on the screen
+
+    ZJS_VALIDATE_ARGS(Z_NUMBER, Z_NUMBER, Z_STRING, Z_ARRAY, Z_OPTIONAL Z_NUMBER);
+    ZJS_GET_HANDLE(this, gfx_handle_t, handle, gfx_type_info);
+    gfx_data_t argData;
+    args_to_data(&argData, argc, argv);
+    return zjs_gfx_draw_char_priv(argData.coords[0], argData.coords[1], argData.text[0], argData.color, argData.size, handle);
+}
+
+static ZJS_DECL_FUNC(zjs_gfx_draw_string)
+{
+    // requires: Requires 4 arguments
+    //           arg[0] - x coord of the top left.
+    //           arg[1] - y coord of the top left.
+    //           arg[2] - text.
+    //           arg[3] - color array.
+    //           arg[4] - optional size of character.
+    //
+    //  effects: Draws a character on the screen
+
+    ZJS_VALIDATE_ARGS(Z_NUMBER, Z_NUMBER, Z_STRING, Z_ARRAY, Z_OPTIONAL Z_NUMBER);
+    ZJS_GET_HANDLE(this, gfx_handle_t, handle, gfx_type_info);
+    gfx_data_t argData;
+    args_to_data(&argData, argc, argv);
+    jerry_value_t ret = ZJS_UNDEFINED;
+
+    for (u8_t i = 0; i < argData.textSize; i++) {
+        ret = zjs_gfx_draw_char_priv(argData.coords[0] + (i * 7 * argData.size), argData.coords[1], argData.text[i], argData.color, argData.size, handle);
```

Comment:
Since you chose a proportional width font, you really should space the characters out based on character width rather than giving them all the same size; it will look a lot better.

I think you can do that pretty easily here by just returning the character width from draw_char_priv and add it to your running x count instead of this i * 7 * size.

Reply:
I've updated this to be smarter.